### ISSUE TITLE
Function to get absolute path to file

### ISSUE THREAD

**Issue by Xpl0itU** [Raudius/files_scripts#30]
When running certain commands from shell I need to pass the full path to the selected file, I made a workaround but it doesn't work with group folders, maybe a full_path_unlocked() function?

**Comment by alpapan** [comment]
Hello

I was going to request the same thing!

At the moment I'm using a Flow external scripts app to get that (denoted as `%f` in their app) but it's clunky as it needs to a trigger.

I was going to find a way to implement it but I couldn't figure out the inheritance, many of the File or Folder API commands where not inherited even though I put the use lines (sorry on mobile right now).

If there a short primer on how to expose functions, I could try and make a pull request.

**Comment by ernolf** [comment]
> .. but it doesn't work with group folders, maybe a full_path_unlocked() function?

You really only need the `etag` provided by `meta_data()`. With that, you can query the filepath from database.
Here an example with a wrapper script. For the sake of simplixcity, I prosume that your webserver-user is www-data and your database mysql on the same host as the nextcloud instance but this can be done with any database and the database credentials can of course be detected automaticaly to make it a more generic usable solution.

In this action-script I pass the etag of selected files to a script `"/usr/local/bin/runscript_logger"` and write stdout and stderr of that script in a text file in the same directory with the same name as the selected files but with an extra appended `.txt`

You can Copy&Paste it into your "File actions" app and give it a name like "Get absolute filepath":
```
local files = get_input_files()

-- Loop through input files
for _, input_file in ipairs(files) do
  local output_filename = input_file.name .. '.txt'
  local out_folder = get_parent(input_file)

  -- Check if the output file already exists
  if exists(out_folder, output_filename) then
    print('Output file "' .. output_filename .. '" already exists. Skipping...')
    -- Skip to the next iteration of the loop
    goto continue
  end

  local arguments = {}
  if is_file(input_file) then
    local metadata = meta_data(input_file)
    local etag = metadata.etag or ""
    table.insert(arguments, '"' .. etag .. '"')
  end

  -- Construct the command string with quotes around each file name
  local command = "/usr/local/bin/runscript-logger" .. table.concat(arguments, " ")

  -- Run the command
  local result = shell_command(command)

  -- Create output file with the same name as the input file but with .txt extension
  local output_file = new_file(out_folder, output_filename, result.output .. "\n" .. result.errors)
  if output_file == nil then
    abort("Could not create output file: " .. output_filename)
  end
  ::continue::
end
```

To log the output of the wrapper-script, create, if not yet exists, a log-directory owned by the webserver-user:
```
mkdir -p /var/log/nextcloud
touch /var/log/nextcloud/runscript.log
chown -R www-data.www-data /var/log/nextcloud
```
In the course of testing, you can now also use
```
tail -f /var/log/nextcloud/runscript.log
```
to monitor stdout and stderr in real time.

Now create the file `/usr/local/bin/runscriptlogger`, which is a basic test script which you can use for better understanding how arguments work, e.g. if you want to pass arguments containing spaces:

In the second part of the script I appended the code for returning the filepath, wich prosumes that the only argument given ($1) is the etag.
You must add your database-user and database password in the variables as shown below and of course make it executable.

`/usr/local/bin/runscriptlogger`:
```
#!/bin/bash

## Basic Part

# log output.
exec > >(tee -a /var/log/nextcloud/runscript.log) 2>&1

echo "$(date +"%Y-%m-%d %H:%M:%S %Z")"

if [ $# -lt 10 ]; then
  digits=1
elif [ $# -lt 100 ]; then
  digits=2
else
  digits=3
fi

# iterate over the arguments
for (( i=0; i<=$#; i++ )); do
  # format the index with leading zeros
  index=$(printf "%0*d" $digits $i)
  # get the i-th argument
  arg=${!i}
  # echo the argument with its index
  echo "  -  arg $index = $arg"
done
echo

## Part 2:
# create an absolute filepath,  $1 has to be etag

# change to your needs:
db_user=**DATABASEUSER**
db_pass=**DATABASEPASSWORD**
db_name=nextcloud
db_pref=oc_
f_etag=$1

db_strg="mysql -u $db_user -p$db_pass --disable-auto-rehash $db_name -srNe"

query_storage() {
    column_name=$1
    shift
    $db_strg "SELECT id FROM ${db_pref}storages WHERE numeric_id = (SELECT storage FROM ${db_pref}filecache WHERE $column_name = '$*')"
}
q_filename="$($db_strg "SELECT path FROM ${db_pref}filecache WHERE etag = '$f_etag'")"
q_storage="$(query_storage etag $f_etag)"

if echo "$q_storage" | grep -Eq ^local::.* ;then
    absolute_filepath="${q_storage#*::}$q_filename"
elif echo "$q_storage" | grep -Eq ^home::.* ;then
    nc_user="${q_storage#*::}"
    q_storage="$(query_storage path $nc_user)"
    absolute_filepath="${q_storage#*::}$nc_user/$q_filename"
else
    echo "storage '$qstorage' not yet suportetd by this script"
fi

echo "$absolute_filepath"

exit 0
```

This works with personal as well as with group folders and local storages.

Please be kind to me, I've only been dealing with *lua* for the first time since yesterday, so for experienced lua programmers my script may look a bit amateurish, but it's about the solution with the file path.

I made a file action with this construct that lists the contents of compressed files and img files so I don't have to download them first.

Therefore it only needs some mor mime_type checks (not absolutely necessary) and the packages `p7zip-full` and `p7zip-rar`.
I named the Action "List content by 7z"

Now you only have to append
```
7z l  "$absolute_filepath"
```
to nearly the same  script as the example file-logger script (without the first part etc), but give it another name like `/usr/local/bin/nc-7z-l`

I find it very handy

Please give me your feedback with the appropriate emoji depending on what you think of my solution 😉

Happy hacking

**Comment by ernolf** [comment]
Albeit there is one limitations with this solution:

* If a file is twice in the filesystem, because it is a symlink on the underlying filesystem (harddisk) to another file on the same harddisk, the etag is the same as well, that means, it is not as unique as the file_id, wich is not the same in such a case.

That can happen if you e.g. mount disks from "outside" per USB (for inspection or backup, e.g. your old sytem drive from an outranged PC or so) and those symlinks has been neccesary in that (former) environment.

But that is a very special case.

**Comment by Raudius** [comment]
I have addressed this issue in #43, and will come in the next release. The path will be available via:

```lua
local path = meta_data(file).local_path
```

@ernolf Kudos for figuring out this impressive work-around! The fact that the workaround was this complex is probably a sign that this was a much needed addition :smile:
